### ISSUE TITLE
Mopidy crashes when pidi is enabled

### ISSUE THREAD

**Issue by chileno01** [pimoroni/mopidy-pidi#12]
I updated mopidy-pidi to see if the latest version solved the issue about showing the name of the song being played when streaming internet radios. Since then, mopidy crashes after running for a while when pidi is enabled. It doesn't happen if pidi is disabled.
I changed logging verbosity to 4 in the logging section of mopidy.conf to see what was going on and the only thing I found before the crashing was this:

May  2 19:50:52 localhost mopidy[737]: free(): invalid pointer
May  2 19:50:53 localhost systemd[1]: mopidy.service: Main process exited, code=killed, status=6/ABRT
May  2 19:50:53 localhost systemd[1]: mopidy.service: Failed with result 'signal'.

I'm running:
Mopidy-PiDi 1.0.2
mopidy 3.0.2 
Python 3.7.3
Raspbian GNU/Linux 10 (buster) 4.19.97+ #1294
Raspberry Pi Zero W

**Comment by Gadgetoid** [comment]
What version of the `spidev` Python library are you running? There was a problem with a memory leak in spidev which may be what you're encountering, and the PiDi plugin responsible for driving out displays currently does not specify that the fixed spidev is required. (due to the Pi being weird and having a mix of pypi/apt dependencies)

**Comment by chileno01** [comment]
Hello: I'm using spidev 3.4 according to what is reported by "pip3 list"

Regards

**Comment by Gadgetoid** [comment]
Sorry, I haven't had any time to look into this yet but I haven't forgotten about it.

It probably wont help, but it might be worth updating to the latest PiDi release anyway: `sudo pip3 install --upgrade mopidy-pidi`.

**Comment by JayGatsby7** [comment]
> Sorry, I haven't had any time to look into this yet but I haven't forgotten about it.
> 
> It probably wont help, but it might be worth updating to the latest PiDi release anyway: `sudo pip3 install --upgrade mopidy-pidi`.

Mopidy is crashing on me too

**Comment by chileno01** [comment]
Hello:
I did the upgrade but mopidy still crashes when pidi is enabled.
The log still shows the same "free(): invalid pointer" message,

**Comment by minils** [comment]
Having the same problem.
Additional Information:
```
pi@raspberrypi:~ $ sudo mopidyctl deps
Running "/usr/bin/mopidy --config /usr/share/mopidy/conf.d:/etc/mopidy/mopidy.conf deps" as user mopidy
WARNING  [MainThread] mopidy.config Ignoring config section 'mpd' because no matching extension was found
Executable: /usr/bin/mopidy
Platform: Linux-4.19.97+-armv6l-with-debian-10.4
Python: CPython 3.7.3 from /usr/lib/python3.7
Mopidy: 3.0.2 from /usr/lib/python3/dist-packages
Mopidy-Podcast: 3.0.0 from /usr/local/lib/python3.7/dist-packages
  Pykka: 2.0.2 from /usr/lib/python3/dist-packages
  Mopidy: 3.0.2 from /usr/lib/python3/dist-packages
  uritools: 3.0.0 from /usr/local/lib/python3.7/dist-packages
  setuptools: 40.8.0 from /usr/lib/python3/dist-packages
  cachetools: 4.1.0 from /usr/local/lib/python3.7/dist-packages
Mopidy-Party: 1.0.0 from /usr/local/lib/python3.7/dist-packages
  setuptools: 40.8.0 from /usr/lib/python3/dist-packages
  Mopidy: 3.0.2 from /usr/lib/python3/dist-packages
  Pykka: 2.0.2 from /usr/lib/python3/dist-packages
Mopidy-Spotify: 4.0.1 from /usr/lib/python3/dist-packages
Mopidy-TuneIn: 1.0.0 from /usr/local/lib/python3.7/dist-packages
  setuptools: 40.8.0 from /usr/lib/python3/dist-packages
  Pykka: 2.0.2 from /usr/lib/python3/dist-packages
  Mopidy: 3.0.2 from /usr/lib/python3/dist-packages
  requests: 2.21.0 from /usr/lib/python3/dist-packages
Mopidy-Iris: 3.49.0 from /usr/local/lib/python3.7/dist-packages
  setuptools: 40.8.0 from /usr/lib/python3/dist-packages
  Mopidy: 3.0.2 from /usr/lib/python3/dist-packages
  Pykka: 2.0.2 from /usr/lib/python3/dist-packages
mopidy-pidi: 1.0.3 from /usr/local/lib/python3.7/dist-packages
  musicbrainzngs: 0.7.1 from /usr/local/lib/python3.7/dist-packages
  Pykka: 2.0.2 from /usr/lib/python3/dist-packages
  setuptools: 40.8.0 from /usr/lib/python3/dist-packages
  netifaces: 0.10.9 from /usr/local/lib/python3.7/dist-packages
  Mopidy: 3.0.2 from /usr/lib/python3/dist-packages
mopidy-raspberry-gpio: 1.0.0 from /usr/local/lib/python3.7/dist-packages
  Mopidy: 3.0.2 from /usr/lib/python3/dist-packages
  setuptools: 40.8.0 from /usr/lib/python3/dist-packages
  Pykka: 2.0.2 from /usr/lib/python3/dist-packages
GStreamer: 1.14.4.0 from /usr/lib/python3/dist-packages/gi
  Detailed information:
    Python wrapper: python-gi 3.30.4
    Relevant elements:
      Found:
        uridecodebin
        souphttpsrc
        appsrc
        alsasink
        osssink
        oss4sink
        pulsesink
        id3demux
        id3v2mux
        lamemp3enc
        mpegaudioparse
        mpg123audiodec
        vorbisdec
        vorbisenc
        vorbisparse
        oggdemux
        oggmux
        oggparse
        flacdec
        flacparse
        shout2send
      Not found:
        flump3dec
        mad
```
Output of `journalctl -u mopidy.service`
```
May 25 07:48:02 raspberrypi systemd[1]: Starting Mopidy music server...
May 25 07:48:02 raspberrypi systemd[1]: Started Mopidy music server.
May 25 07:48:13 raspberrypi mopidy[3128]: INFO     [MainThread] mopidy.__main__ Starting Mopidy 3.0.2
May 25 07:48:13 raspberrypi mopidy[3128]: INFO     [MainThread] mopidy.config Loading config from builtin defaults
May 25 07:48:13 raspberrypi mopidy[3128]: INFO     [MainThread] mopidy.config Loading config from file:///usr/share/mopidy/conf.d/mopidy.conf
May 25 07:48:13 raspberrypi mopidy[3128]: INFO     [MainThread] mopidy.config Loading config from file:///etc/mopidy/mopidy.conf
May 25 07:48:13 raspberrypi mopidy[3128]: INFO     [MainThread] mopidy.config Loading config from command line options
May 25 07:48:13 raspberrypi mopidy[3128]: WARNING  [MainThread] mopidy.config Ignoring config section 'mpd' because no matching extension was found
May 25 07:48:17 raspberrypi mopidy[3128]: INFO     [MainThread] mopidy.__main__ Enabled extensions: file, pidi, tunein, http, spotify, iris, m3u, raspberry-gpio, softwaremixer, party, stream
May 25 07:48:17 raspberrypi mopidy[3128]: INFO     [MainThread] mopidy.__main__ Disabled extensions: none
May 25 07:48:19 raspberrypi mopidy[3128]: INFO     [MainThread] mopidy.commands Starting Mopidy mixer: SoftwareMixer
May 25 07:48:19 raspberrypi mopidy[3128]: INFO     [MainThread] mopidy.commands Mixer volume set to 40
May 25 07:48:19 raspberrypi mopidy[3128]: INFO     [MainThread] mopidy.commands Starting Mopidy audio
May 25 07:48:19 raspberrypi mopidy[3128]: INFO     [MainThread] mopidy.commands Starting Mopidy backends: TuneInBackend, FileBackend, M3UBackend, StreamBackend, SpotifyBackend
May 25 07:48:19 raspberrypi mopidy[3128]: INFO     [Audio-2] mopidy.audio.actor Audio output set to "alsasink device=hw:sndrpihifiberry"
May 25 07:48:22 raspberrypi mopidy[3128]: INFO     [SpotifyEventLoop] mopidy_spotify.backend Logged in to Spotify in online mode
May 25 07:48:22 raspberrypi mopidy[3128]: INFO     [SpotifyBackend-7] mopidy_spotify.web Logged into Spotify Web API as ***********
May 25 07:49:21 raspberrypi mopidy[3128]: INFO     [SpotifyBackend-7] mopidy_spotify.playlists Refreshed 56 Spotify playlists
May 25 07:49:21 raspberrypi mopidy[3128]: INFO     [MainThread] mopidy.commands Starting Mopidy core
May 25 07:49:22 raspberrypi mopidy[3128]: INFO     [MainThread] mopidy.commands Starting Mopidy frontends: RaspberryGPIOFrontend, PiDiFrontend, IrisFrontend, HttpFrontend
May 25 07:49:22 raspberrypi mopidy[3128]: INFO     [IrisFrontend-14] mopidy_iris.core Starting Iris 3.49.0
May 25 07:49:22 raspberrypi mopidy[3128]: INFO     [HttpFrontend-16] mopidy.http.actor HTTP server running at [::ffff:0.0.0.0]:6680
May 25 07:49:22 raspberrypi mopidy[3128]: INFO     [MainThread] mopidy.commands Starting GLib mainloop
May 25 10:28:49 raspberrypi mopidy[3128]: free(): invalid pointer
May 25 10:28:49 raspberrypi systemd[1]: mopidy.service: Main process exited, code=killed, status=6/ABRT
May 25 10:28:49 raspberrypi systemd[1]: mopidy.service: Failed with result 'signal'.
```
Just installed a fresh raspbian. Plugged in the pirate audio hat and ran the install script.

The error disappears after disabling the pidi module.

**Comment by Gadgetoid** [comment]
Currently in the process of trying to replicate this- I strongly suspect it’s either spidev (again) or PIL but I’ll admit that I’m not the most competent at tracing these kind of buried problems.

Some more information might be helpful if you have time:
1. Raspberry Pi model
2. RAM amount

Will update as I think of more possible distinguishing factors.

As far as I’m aware Mopidy-PiDi was stable after the first spidev fix, so I’ll check recent dependency updates also.

**Comment by minils** [comment]
Thanks for investigating!
1. Pi Model:
```
pi@raspberrypi:~ $ cat /proc/cpuinfo 
processor	: 0
model name	: ARMv6-compatible processor rev 7 (v6l)
BogoMIPS	: 997.08
Features	: half thumb fastmult vfp edsp java tls 
CPU implementer	: 0x41
CPU architecture: 7
CPU variant	: 0x0
CPU part	: 0xb76
CPU revision	: 7

Hardware	: BCM2835
Revision	: 900093
Serial		: 00000000aab60a88
Model		: Raspberry Pi Zero Rev 1.3
```
2. RAM amount:
```
pi@raspberrypi:~ $ free
              total        used        free      shared  buff/cache   available
Mem:         442820       87028      244160        3144      111632      302172
Swap:        102396           0      102396
```

**Comment by Gadgetoid** [comment]
Thank you- right now I'm running a couple of tests to see if I can isolate where the failure is happening or, indeed, recreate it at all.

Currently run over half a million in-test display updates with the latest PIL/Pillow with no sign of failure. 

May be worth knowing the output of `python3 -m pip freeze | grep Pillow` and `python3 -m pip freeze | grep spidev`

@JayGatsby7 and @chileno01 are you also on Pi Zeros?

**Comment by minils** [comment]
```
pi@raspberrypi:~ $ python3 -m pip freeze | grep Pillow
Pillow==5.4.1
pi@raspberrypi:~ $ python3 -m pip freeze | grep spidev
spidev==3.4
```
As far as I noticed, the error only occurs after several minutes/hours.

Edit: Another thing I noticed is that the error does not occure if no playback is started (after a reboot).

**Comment by chileno01** [comment]
Hello.
Here's my information:
$python3 -m pip freeze | grep Pillow
Pillow==5.4.1
$python3 -m pip freeze | grep spidev
spidev==3.4
$cat /proc/cpuinfo
processor	: 0
model name	: ARMv6-compatible processor rev 7 (v6l)
BogoMIPS	: 697.95
Features	: half thumb fastmult vfp edsp java tls
CPU implementer: 0x41
CPU architecture: 7
CPU variant	: 0x0
CPU part	: 0xb76
CPU revision	: 7

Hardware	: BCM2835
Revision	: 900093
Serial	        : 000000005058189a
Model	        : Raspberry Pi Zero Rev 1.3
`

**Comment by Gadgetoid** [comment]
I'm strongly suspecting some sort of memory leak, probably relating to PIL, but that's a wild guess. However since PiDi itself has not changed, it seems fair to point at a dependency.

It might be worth running `sudo python3 -m pip install --upgrade Pillow`, which should install `Pillow==7.1.2` and seeing if it has any impact.

**Comment by Gadgetoid** [comment]
Okay my SPI test just failed at just over 1 million writes with- you guessed it- `free(): invalid pointer`.

I'm confused, because when I first looked into the SPI issues I used pretty much this same test to verify that the spidev library update fixed the problem. Either I failed to test sufficiently beforehand, or something has regressed since that is causing this problem.

**Comment by Gadgetoid** [comment]
Oh boy I wish I had looked here first: https://github.com/doceme/py-spidev/issues/91

It looks like the "free(): invalid pointer" issue is *not* the same as the general "segmentation fault" I was encountering in my tests around January. This could easily be a kernel regression.

I'll have to put together a pure-C ioctl test to see if I can recreate the problem outside of Python.

**Comment by Gadgetoid** [comment]
To keep you folks in the loop, this is where I'm at right now: https://github.com/doceme/py-spidev/pull/93

I have identified a reference counting bug in the Python C extension that handles SPI communication. It basically meant that the reference counter would overflow at 2^32 and... well... this would happen.

Hopefully I'll get this pushed through to release quickly, but if you want to try the fix out you can:

```
git clone https://github.com/Gadgetoid/py-spidev -b patch-rxbuf-memory-leak
cd py-spidev
python3 setup.py build
sudo python3 setup.py install
```

Which should- in theory- replace your version of spidev with the fixed version. It will have the same version number - 3.4 - so should be replaced when the real deal is released.

**Comment by JayGatsby7** [comment]
> To keep you folks in the loop, this is where I'm at right now: [doceme/py-spidev#93](https://github.com/doceme/py-spidev/pull/93)
> 
> I have identified a reference counting bug in the Python C extension that handles SPI communication. It basically meant that the reference counter would overflow at 2^32 and... well... this would happen.
> 
> Hopefully I'll get this pushed through to release quickly, but if you want to try the fix out you can:
> 
> ```
> git clone https://github.com/Gadgetoid/py-spidev -b patch-rxbuf-memory-leak
> cd py-spidev
> python3 setup.py build
> sudo python3 setup.py install
> ```
> 
> Which should- in theory- replace your version of spidev with the fixed version. It will have the same version number - 3.4 - so should be replaced when the real deal is released.

trying this now.  Thank you!

**Comment by chileno01** [comment]
> To keep you folks in the loop, this is where I'm at right now: [doceme/py-spidev#93](https://github.com/doceme/py-spidev/pull/93)
> 
> I have identified a reference counting bug in the Python C extension that handles SPI communication. It basically meant that the reference counter would overflow at 2^32 and... well... this would happen.
> 
> Hopefully I'll get this pushed through to release quickly, but if you want to try the fix out you can:
> 
> ```
> git clone https://github.com/Gadgetoid/py-spidev -b patch-rxbuf-memory-leak
> cd py-spidev
> python3 setup.py build
> sudo python3 setup.py install
> ```
> 
> Which should- in theory- replace your version of spidev with the fixed version. It will have the same version number - 3.4 - so should be replaced when the real deal is released.

I followed your instructions and apparently the new version of py-spidev was installed.
I reactivated pidi but after 4 hours of running, mopidy crashed with the same free(): invalid pointer message

**Comment by JayGatsby7** [comment]
> > To keep you folks in the loop, this is where I'm at right now: [doceme/py-spidev#93](https://github.com/doceme/py-spidev/pull/93)
> > I have identified a reference counting bug in the Python C extension that handles SPI communication. It basically meant that the reference counter would overflow at 2^32 and... well... this would happen.
> > Hopefully I'll get this pushed through to release quickly, but if you want to try the fix out you can:
> > ```
> > git clone https://github.com/Gadgetoid/py-spidev -b patch-rxbuf-memory-leak
> > cd py-spidev
> > python3 setup.py build
> > sudo python3 setup.py install
> > ```
> > 
> > 
> > Which should- in theory- replace your version of spidev with the fixed version. It will have the same version number - 3.4 - so should be replaced when the real deal is released.
> 
> I followed your instructions and apparently the new version of py-spidev was installed.
> I reactivated pidi but after 4 hours of running, mopidy crashed with the same free(): invalid pointer message

Mine has been stable for a few days with the fix.  I will let you know if I experience further issues.

**Comment by Gadgetoid** [comment]
@chileno01 might be worth trying to uninstall all versions of the library first and try again.

If you run the following a few times it should uninstall any instances of spidev:

```python
sudo python3 -m pip uninstall spidev
```

And additionally `sudo apt remove python3-spidev`

**Comment by chileno01** [comment]
> @chileno01 might be worth trying to uninstall all versions of the library first and try again.
> 
> If you run the following a few times it should uninstall any instances of spidev:
> 
> ```python
> sudo python3 -m pip uninstall spidev
> ```
> 
> And additionally `sudo apt remove python3-spidev`

Hello @Gadgetoid : Thanks for the tip, but the first command is giving me
`pi@mopidy:~ $ sudo python3 -m pip uninstall spidev
Cannot uninstall 'spidev'. It is a distutils installed project and thus we cannot accurately determine which files belong to it which would lead to only a partial uninstall.
`
Any ideas?

**Comment by minils** [comment]
> To keep you folks in the loop, this is where I'm at right now: [doceme/py-spidev#93](https://github.com/doceme/py-spidev/pull/93)
> 
> I have identified a reference counting bug in the Python C extension that handles SPI communication. It basically meant that the reference counter would overflow at 2^32 and... well... this would happen.
> 
> Hopefully I'll get this pushed through to release quickly, but if you want to try the fix out you can:
> 
> ```
> git clone https://github.com/Gadgetoid/py-spidev -b patch-rxbuf-memory-leak
> cd py-spidev
> python3 setup.py build
> sudo python3 setup.py install
> ```
> 
> Which should- in theory- replace your version of spidev with the fixed version. It will have the same version number - 3.4 - so should be replaced when the real deal is released.

I applied the fix last week and had no crashes so far...

**Comment by Gadgetoid** [comment]
@chileno01 iirc that means you should run `sudo apt remove python3-spidev` but it's been a while since I've run into that problem.

**Comment by chileno01** [comment]
> @chileno01 iirc that means you should run `sudo apt remove python3-spidev` but it's been a while since I've run into that problem.

So far so good. I have been listening to radio streams for 24 hours and haven't had crashes so far.
And the display is showing the name of the song being streamed, that was the original request.
Thanks for all your help and dedication

**Comment by minils** [comment]
Also got no errors since I applied the fix.
I think this is resolved.
Thanks a lot for getting into it! 😎

**Comment by rhmk** [comment]
Thanks for documenting it. Just bought a pirate audio device and ran in the exact same bug. I removed the spidev version 3.4 from the distribution (apt remove spidev) and installed version 3.5 ( python3 -m pip install spidev  ). My problem seems to be gone, although the above-mentioned fix has not made it into 3.5, but a lot of others, that might relate to it.

**Comment by rhmk** [comment]
My package list
```
# pip3 list
Package                   Version
------------------------- ---------
asn1crypto                0.24.0
cachetools                3.1.0
certifi                   2018.8.24
chardet                   3.0.4
cryptography              2.6.1
entrypoints               0.3
font-roboto               0.0.1
fonts                     0.0.3
idna                      2.6
keyring                   17.1.1
keyrings.alt              3.1.1
Mopidy                    3.0.2
Mopidy-Cd                 0.5
Mopidy-Local              3.1.1
Mopidy-MPD                3.0.0
Mopidy-MusicBox-Webclient 3.1.0
mopidy-pidi               1.0.3
Mopidy-Podcast            3.0.0
mopidy-raspberry-gpio     1.0.2
Mopidy-Spotify            4.0.1
Mopidy-TuneIn             1.0.0
musicbrainzngs            0.6
netifaces                 0.10.9
numpy                     1.16.2
olefile                   0.46
pidi-display-pil          0.1.0
pidi-display-st7789       0.1.1
Pillow                    5.4.1
pip                       20.2.2
pipdeptree                1.0.0
pycrypto                  2.6.1
PyGObject                 3.30.4
Pykka                     2.0.2
pyspotify                 2.1.3
python-apt                1.8.4.1
pyxdg                     0.25
requests                  2.21.0
RPi.GPIO                  0.7.0
SecretStorage             2.3.1
setuptools                40.8.0
six                       1.12.0
spidev                    3.5
ssh-import-id             5.7
ST7789                    0.0.2
tornado                   5.1.1
uritools                  2.2.0
urllib3                   1.24.1
wheel                     0.32.3
```
